Remora.Discord.Extensions/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,97 @@
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Extensions.DependencyInjection;
+using Remora.Commands.Extensions;
+using Remora.Commands.Groups;
+using Remora.Discord.Extensions.Attributes;
+using Remora.Discord.Gateway.Extensions;
+using Remora.Discord.Gateway.Responders;
+
+namespace Remora.Discord.Extensions.Extensions;
+
+/// <summary>
+/// A collection of extensions for the <see cref="IServiceCollection"/> interface.
+/// </summary>
+public static class ServiceCollectionExtensions
+{
+    /// <summary>
+    /// Automatically registers every command group in the given assembly.
+    /// </summary>
+    /// <param name="serviceCollection">The service collection to add commands to.</param>
+    /// <param name="assembly">The assembly to discover command groups from.</param>
+    /// <param name="treeName">The name of the tree to register commands to, otherwise the default.</param>
+    /// <param name="typeFilter">A function to select whether a given command group should be registered.</param>
+    /// <returns>The service collection with registered commands.</returns>
+    public static IServiceCollection AddDiscordCommands
+    (
+        this IServiceCollection serviceCollection,
+        Assembly assembly,
+        string? treeName = null,
+        Func<Type, bool>? typeFilter = null
+    )
+    {
+        var canidates = assembly.GetTypes()
+                                .Where(t => t.IsClass && !t.IsAbstract && typeof(CommandGroup).IsAssignableFrom(t))
+                                .ToArray();
+
+        var tree = serviceCollection.AddCommandTree(treeName);
+
+        foreach (var canidate in canidates)
+        {
+            if (typeFilter?.Invoke(canidate) ?? true)
+            {
+                tree.WithCommandGroup(canidate);
+            }
+        }
+
+        return serviceCollection;
+    }
+
+    /// <summary>
+    /// Adds all responders in the given assembly to the service collection, using their attributed group if possible.
+    /// </summary>
+    /// <param name="serviceCollection">The service collection to register responders in.</param>
+    /// <param name="assembly">The assembly to discover responders from.</param>
+    /// <returns>The service collection to chain calls.</returns>
+    public static IServiceCollection AddResponders(this IServiceCollection serviceCollection, Assembly assembly)
+    {
+        var canidates = assembly.GetTypes()
+                                .Where(t => t.IsClass &&
+                                           !t.IsAbstract &&
+                                            t.GetInterfaces() // .Any(t => typeof(IResponder).IsAssignableFrom(t)) ?
```

Comment:
I presume the comment isn't supposed to be here?

Reply:
There's a question mark because I thought of it while I was writing it, but unsure of if I should use it

Reply:
@Nihlus for input here?